File: 1979. Find Greatest Common Divisor of Array/1979.cpp
#include <iostream>
#include <vector>
#include <algorithm> //support minmax_element()
using namespace std;

class Solution {
public:
    int findGCD_1(vector<int>& nums) {
        /* find max & min */
        auto result = minmax_element(nums.begin(), nums.end());
        int min_num = *(result.first);
        int max_num = *(result.second);
        /* calculate GCD: Euclidean Algorithm */
        int a = max_num;
        int b = min_num;
        while (b != 0) {
            int tmp = b;
            b = a % b;
            a = tmp;
        }
        return a;
    }
    int findGCD_2(vector<int>& nums){
        /* find max & min */
        int max = nums[0], min = nums[0];
        for (auto &num :nums){
            if (num > max)  max = num;
            else if (num < min)  min = num;
        }
        /* calculate GCD: Euclidean Algorithm */
        int a = max;
        int b = min;
        while (b != 0) {
            int tmp = b;
            b = a % b;
            a = tmp;
        }
        return a;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {2,5,6,9,10};
    cout << sol.findGCD_2(nums) << endl;
    return 0;
}

//compile: g++ -std=c++11 1979.cpp -o 1979
//execute: ./1979

File: easy/1979. Find Greatest Common Divisor of Array/1979.py
class Solution(object):
    def findGCD(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        min_num = min(nums)
        max_num = max(nums)
        if (max_num%min_num == 0):
            return min_num
        """ Euclidean Algorithm """
        a = max_num
        b = min_num
        while (b != 0):
            tmp = b
            b = a % b
            a = tmp
        return a

if __name__ == '__main__':
    nums = [2,5,6,9,10]
    solution = Solution()
    print(solution.findGCD(nums))

#python3 1979.py